web/packages/shared/components/ButtonFileUpload/ButtonFileUpload.tsx

```diff
@@ -0,0 +1,130 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useEffect, useState, useRef } from 'react';
+import styled from 'styled-components';
+import { Upload } from 'design/Icon';
+import { ButtonBorder, Flex, LabelInput } from 'design';
+
+import type { ButtonSize } from 'design/Button';
+import type { IconSize } from 'design/Icon/Icon';
+
+/**
+ * ButtonFileUpload let the user choose a file from local filesystem.
+ * Only single file can be selected. To support multiple file uploads,
+ * update this component with `multiple` attribute.
+ * @param text button text.
+ * @param errorMessage error message to display when a file is not selected.
+ * @param showValidationError display errorMessage on file validation error.
+ * @param accept whitelist file extension for the file picker.
+ * @param disabled enable or disable button.
+ * @param onFileSelect callback function to process selected file.
+ * @param buttonSize button display size.
+ * @param buttonIconSize upload icon display size.
+ */
+export function ButtonFileUpload({
+  text,
+  errorMessage,
+  accept,
+  disabled,
+  onFileSelect,
+  showValidationError,
+  buttonSize = 'medium',
+  buttonIconSize = 'medium',
+}: {
+  text: string;
+  errorMessage: string;
+  accept: string;
+  disabled: boolean;
+  onFileSelect: (file: File) => void;
+  showValidationError?: boolean;
+  buttonSize?: ButtonSize;
+  buttonIconSize?: IconSize;
+}) {
+  const fileInputRef = useRef<HTMLInputElement>();
+  const [fileInputError, setFileInputError] =
+    useState<boolean>(showValidationError);
+  if (showValidationError !== fileInputError) {
+    if (!fileInputError) {
+      setFileInputError(showValidationError);
+    }
+  }
```

Comment:
hrm, i'm not sure i understand what this is for? should it go in a use effect hook where `showValidatiionError` changes?

Reply:
Oh this is just my attempt at [not using an effect when we need to adjust some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes), which is I have been pointed at couple times now 😃 This naturally fits into the `useEffect` hook but is currently implemented exactly according to the guide to remove usage of effect (check the exact link). 

The `if (showValidationError !== fileInputError) {` condition is necessary to avoid loop where pops is used to update state outside of effect. 
And inside the if condition, we only want to update `fileInputError` with value from `showValidationError` only if the `fileInputError` is `false` or `null`.

Reply:
ahh ok, i need to apparently learn this too 😅 thanks for the link